safedelete/models.py

```diff
@@ -278,3 +283,37 @@ def __init__(self, *args, **kwargs):
         warnings.warn('The SafeDeleteMixin class was renamed SafeDeleteModel',
                       DeprecationWarning)
         SafeDeleteModel.__init__(self, *args, **kwargs)
+
+
+class SafeDeleteCascadeControlModel(SafeDeleteModel):
+    """ Abstract safedelete-ready model with additional option to control cascade undelete.
+
+    .. note::
+        This abstract class inherits from SafeDeleteModel, so it has the same usability. However
+        this class implements an additional attribute called by default as `has_cascaded` that
+        allows undelete cascading to restore only child classes that were also deleted by a
+        cascading operation, i.e. all objects that were deleted before their parent object deletion,
+        should keep deleted if the same parent object is restored by undelete method.
+
+    :attribute has_cascaded:
+        BooleanField set True whenever the object is deleted due cascade operation called by delete
+        method of any parent Model. Default value is False. Later if its parent model calls for
+        undelete, cascading undelete will consider only the ones that were set to True before.
+    """
+
+    def soft_delete_policy_action(self, **kwargs):
+        if kwargs.get('is_cascade'):
+            setattr(self, HAS_CASCADED_FIELD_NAME, True)
+        super(SafeDeleteCascadeControlModel, self).soft_delete_policy_action(**kwargs)
+
+    def save(self, keep_deleted=False, **kwargs):
+        if not keep_deleted:
+            if getattr(self, HAS_CASCADED_FIELD_NAME, False):
```

Comment:
This line can be removed. We always end up having the attribute at False afterwards.